src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtAbstractProver.java

```diff
@@ -0,0 +1,278 @@
+    osmtSolver = new MainSolver(creator.getEnv(), pConfig, "JavaSmt");
+
+    assertionStack.push(new ArrayList<>()); // create initial level
+  }
+
+  protected static SMTConfig getConfigInstance(int randomSeed, boolean interpolation, int algBool, int algUf, int algLra) {
+    SMTConfig config = new SMTConfig();
+    config.setRandomSeed(randomSeed);
+    config.setInterpolation(interpolation);
+
+    if (interpolation) {
+      config.setOption(":interpolation-bool-algorithm", new SMTOption(algBool));
+      config.setOption(":interpolation-euf-algorithm", new SMTOption(algUf));
+      config.setOption(":interpolation-lra-algorithm", new SMTOption(algLra));
+    }
+    return config;
+  }
+
+  final MainSolver getOsmtSolver() {
+    return osmtSolver;
+  }
+
+  @Override
+  public void push() {
+    Preconditions.checkState(!closed);
+    setChanged();
+    assertionStack.push(new ArrayList<>());
+    osmtSolver.push();
+  }
+
+  @Override
+  public void pop() {
+    Preconditions.checkState(!closed);
+    setChanged();
+    Preconditions.checkState(size() > 0, "Tried to pop from an empty solver stack");
+    assertionStack.pop();
+    osmtSolver.pop();
+  }
+
+
+  @Nullable
+  protected abstract T addConstraintImpl(PTRef f) throws InterruptedException;
+  
+  @Override
+  @Nullable
+  public T addConstraint(BooleanFormula pF) throws InterruptedException {
+    Preconditions.checkState(!closed);
+    setChanged();
+
+    PTRef f = creator.extractInfo(pF);
+    T label = addConstraintImpl(f);
+
+    assertionStack.peek().add(f);
+    return label;
+  }
+
+  @SuppressWarnings("resource")
+  @Override
+  public Model getModel() {
+    Preconditions.checkState(!closed);
+    checkGenerateModels();
+
+    Model model = new OpenSmtModel(this, creator, getAssertedFormulas());
+    return registerEvaluator(model);
+  }
+
+  @Override
+  public Evaluator getEvaluator() {
+    Preconditions.checkState(!closed);
+    checkGenerateModels();
+    return getEvaluatorWithoutChecks();
+  }
+
+  @SuppressWarnings("resource")
+  @Override
+  protected Evaluator getEvaluatorWithoutChecks() {
+    return registerEvaluator(new OpenSmtEvaluator(this, creator));
+  }
+
+  protected void setChanged() {
+    if (!changedSinceLastSatQuery) {
+      changedSinceLastSatQuery = true;
+      closeAllEvaluators();
+    }
+  }
+
+  @Override
+  public ImmutableList<ValueAssignment> getModelAssignments() throws SolverException {
+    Preconditions.checkState(!closed);
+    Preconditions.checkState(!changedSinceLastSatQuery);
+    return super.getModelAssignments();
+  }
+
+  /** Make sure that the assertions only use features supported by the selected sublogic. */
+  private void checkCompatability() throws SolverException {
+
+    Logic osmtLogic = creator.getEnv();
+    PTRef asserted = osmtLogic.mkAnd(new VectorPTRef(getAssertedFormulas()));
```

Comment:
The conjunction here is dangerous, because a solver could internally simplify the assertion and remove symbols from the result.